### PR TITLE
Npu adapt megatron

### PR DESCRIPTION
**Pull request by addsubmuldiv**
# PR Type
- [x] Bug Fix
- [x] New Feature
- [x] Document Updates
- [ ] More Models or Datasets Support

# Summary

This PR completes Twinkle's NPU Megatron adaptation and targets the Twinkle + Megatron-LM 0.15.3 + MindSpeed 0.15.3 + mcore-bridge stack. The goal is to make the dense / LoRA 8-card training path stable on NPU.

Main changes:
- Move MindSpeed bootstrap before `mcore_bridge` is imported to avoid late patching and early binding of TE / Megatron symbols.
- Build MindSpeed runtime args from the current `ModelConfig` and the runtime parallel topology, then call `repatch()` when the runtime signature changes.
- Fix distributed initialization and metric gathering on NPU:
  - add a default PG fallback for single-rank local smoke
  - reuse Megatron's Gloo DP group for Python object gathering on NPU
- Fix causal mask handling for NPU FlashAttention:
  - stop feeding Twinkle's 4D dense causal mask directly into the MindSpeed TE flash path
  - let MindSpeed generate the compressed causal mask on the causal NPU path
- Complete multi-LoRA compatibility for the NPU Megatron path:
  - multi-tenant LoRA training
  - multi-tenant save/export flow
  - optimizer capability selection cleanup

# What Changed

## 1. MindSpeed runtime bootstrap
- Added an NPU-only runtime bootstrap to ensure MindSpeed patching happens before `mcore_bridge` import.
- Unified MindSpeed runtime arg generation into one path so Twinkle and MindSpeed do not read inconsistent runtime state.

## 2. Process group / metric gather
- Fixed default PG initialization for single-rank Megatron smoke.
- Changed NPU `gather_object()` to prefer Megatron's Gloo DP group to avoid hangs in metrics / Python object gathering.
- Kept the DP+CP group selection for CP-enabled runs.

## 3. NPU FlashAttention
- Fixed causal attention mask handling on NPU.
- For causal NPU paths, no longer pass Twinkle's 4D dense mask directly, avoiding the MindSpeed TE FlashAttention shape mismatch.

## 4. LoRA / Multi-LoRA
- Fixed runtime checks for LoRA finalize so a bare model with `ddp_config` is not incorrectly treated as a model that can run native finalize.
- Cleaned up optimizer capability selection for multi-LoRA so it uses the local bf16 optimizer path that fits the model structure.
- Fixed the multi-LoRA save callback signature so the current tenant adapter is correctly passed through during save.

## 5. Documentation
- Updated the NPU support docs with Megatron backend installation and usage guidance.
- Added installation notes for Megatron / MindSpeed / mcore-bridge and the matching cookbook smoke entrypoints.

# Notes

This PR targets the following version stack:
- Megatron-LM 0.15.3
- MindSpeed 0.15.3
- mcore-bridge
- Twinkle NPU environment

### CODE REVIEW

File: src/twinkle/model/megatron/megatron.py [hunk 1]
```diff
@@ -146,6 +151,32 @@ def __init__(
         self.active_group = _default_adapter_name
         MegatronPeft().__call__()
 
+    def _ensure_megatron_process_group(self):
```
Comment (tastelikefeet) [inline src/twinkle/model/megatron/megatron.py]: 上面的self._try_init_process_group() 应该包含了process_group的初始化操作，为什么需要处理第二次呢

File: src/twinkle/model/megatron/megatron.py [hunk 2]
```diff
@@ -358,6 +389,26 @@ def post_loss_function(output_tensor, inputs, logps):
...
+            # processor, however, always expands the 1D token mask into a 4D
+            # dense causal mask. On NPU this makes FlashAttention receive the
+            # wrong mask shape and the real 8-card run fails in
+            # ``aclnnFlashAttentionScore``. For decoder-only causal training
+            # with right padding, the 4D mask is redundant: a causal mask
+            # already prevents valid tokens from attending to the padded tail,
+            # and padded query positions are ignored by labels == -100. So on
+            # the NPU TE path, drop this dense mask and let MindSpeed build the
+            # compressed causal mask it requires.
+            if Platform.device_prefix() == 'npu':
```
Comment (tastelikefeet) [inline src/twinkle/model/megatron/megatron.py]: 这个if放入InputProcessor里面是否更合适

File: src/twinkle/utils/framework.py [hunk 3]
```diff
@@ -42,6 +42,18 @@ def gather_object(object: Any, device_mesh: DeviceMesh, process_group=None):
...
+                # On NPU, letting Python object collectives use the default HCCL
+                # group previously hung in 8-card metric collection at
+                # ``dist.all_gather_object(...)``. Reuse Megatron's dedicated Gloo
+                # DP group instead. When CP is enabled we must pick the DP+CP
+                # variant, otherwise the rank span for metric aggregation is wrong.
+                try:
+                    from megatron.core import parallel_state as mpu
+                    process_group = mpu.get_data_parallel_group_gloo(
+                        with_context_parallel=getattr(device_mesh, 'cp_world_size', 1) > 1)
+                except (ImportError, ModuleNotFoundError):
```
Comment (tastelikefeet) [inline src/twinkle/utils/framework.py]: 这个try except应该没有意义吧 如果初始化出错了就应该抛错

File: src/twinkle/model/megatron/megatron.py (hunk 2 above)
Comment (addsubmuldiv) [reply src/twinkle/model/megatron/megatron.py]: 这个判断依赖 unwrapped_model.config.attention_mask_type，该属性依赖的是 Megatron 运行时的模型配置，只能在 forward 运行时从模型实例上获取。InputProcessor 是纯数据处理组件，不持有模型引用，无法做这个判断。我把这块封装了一下简洁一点

File: src/twinkle/model/megatron/megatron.py (hunk 1 above)
Comment (addsubmuldiv) [reply src/twinkle/model/megatron/megatron.py]: 这块是有点问题，已经重写相关逻辑

File: src/twinkle/utils/framework.py (hunk 3 above)
Comment (addsubmuldiv) [reply src/twinkle/utils/framework.py]: 已按建议改掉

**Review by tastelikefeet — APPROVED** [review]
